wallet-api/test/Spec.hs

```diff
@@ -1,421 +1,469 @@
+              | now >= start + 1 = Annotate "Transaction has been validated" (Boolean $ tx `elem` chain)
+              | otherwise = Top
+              where now   = lastSlot (es^.chainNewestFirst)
+                    chain = concat (es^.chainNewestFirst)
+
+txnUpdateUtxo :: Property
+txnUpdateUtxo = forAllChains (gm ()) $ \es () -> property $
+    length (es^.txPool) >= 1 ==> do
+        tx <- elements $ es^.txPool
+        let slot = lastSlot (es^.chainNewestFirst)
+            ValidatedBlock validTxns events [] _ = validateBlock slot (es^.index) (tx : es^.txPool)
+            noDuplicates = nub validTxns === validTxns
+            duplicatedTransactionFails = not $ null $ flip filter events $
+                \case TxnValidationFail txi _
+                          | txi == hashTx tx -> True
+                          | otherwise -> False
+                      _ -> False
+        pure $ noDuplicates .&&. duplicatedTransactionFails
+
+txnInvalid :: Property
+txnInvalid = forAllChains gm' $ \_ checked -> discardUnless checked
+    where gm' = gm True & userActions .~ [generateAndVerifyValidTransaction]
+          generateAndVerifyValidTransaction mc = do
+              tx <- genInvalidTransaction keys keys mc (gm'^.feeEstimator)
+              pure $ checkAndRemember "Generate an invalid transaction" $ \es ->
+                  (Annotate "Check that the transaction is invalid according to the emulator"
+                   $ Not $ Boolean $ Main.isValid es tx) :&&
+                  (Annotate "Check that the transaction is invalid according to the Mockchain"
+                   $ Not $ Boolean $ Mockchain.isValid (mc^.currentSlot) (mc^.utxo) tx)
+
+slotCorrect :: Property
+slotCorrect = forAllChains gm' $ \_ checked -> discardUnless checked
+    where gm' = gm False & userActions .~ [checkSlot]
+          checkSlot mc = pure $ checkAndRemember "Mockchain slot is identical to emulator slot" $ \es ->
+              Boolean (lastSlot (es^.chainNewestFirst) == mc^.currentSlot)
+
+valueConserved :: Property
+valueConserved = property $ do
+    fee <- choose (0, 10)
+    let gm' = gm () & feeEstimator .~ constantFee (Ada.fromInt fee)
+    pure $ forAllChains gm' $ \es () ->
+        let totalFunds outs   = Map.foldl' Value.plus Value.zero $ txOutValue <$> outs
+            fees              = Ada.adaValueOf $ sum $ Ada.toInt . txFee <$> join (es^.chainNewestFirst)
+            totalInitialFunds = foldl' Value.plus Value.zero (snd <$> gm'^.initialBalances)
+        in property $ totalFunds (getIndex $ es^.index) `Value.plus` fees === totalInitialFunds
+
+outputsConserved :: Property
+outputsConserved = property $ forAllChains (gm ()) $ \es () ->
+    -- Check that every output is either spent or is part of the UTxO set
+    let chain   = join $ es^.chainNewestFirst
+        outs    = foldl' Set.union Set.empty $ Set.fromList . map snd . txOutRefs <$> chain
+        ins     = foldl' Set.union Set.empty $ Set.map txInRef . txInputs <$> chain
```

Comment:
Here and in a few other places you can just use `Data.Foldable.fold`

Reply:
Yeah, my bad. This works for the ones constructing `Set`, of which there's only a couple. On this issue, is there a reason why `Value` is not a monoid? It would let us do the same simplification.

Reply:
Value should probably be a monoid indeed.